### PR TITLE
HIVE-23633: Metastore some JDO query objects do not close …

### PR DESCRIPTION
**Pull request by dengzhhu653**
…properly

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java [hunk 1]
```diff
@@ -914,6 +914,7 @@ private boolean isViewTable(String catName, String dbName, String tblName) throw
     long queryTime = doTrace ? System.nanoTime() : 0;
     MetastoreDirectSqlUtils.timingTrace(doTrace, queryText, start, queryTime);
     if (sqlResult.isEmpty()) {
+      query.closeAll();
       return Collections.emptyList(); // no partitions, bail early.
     }
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:919]: Can you just remove this entire block?  This will allow for a single return statement at the bottom and also allows the cleanup to happen.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java [hunk 2]
```diff
@@ -1453,6 +1454,7 @@ public ColumnStatistics getTableStats(final String catName, final String dbName,
     };
     List<Object[]> list = Batchable.runBatched(batchSize, colNames, b);
     if (list.isEmpty()) {
+      b.closeAllQueries();
       return null;
     }
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:1459]: Same here....

```java
    List<Object[]> list = Batchable.runBatched(batchSize, colNames, b);
    final ColumnStatistics result;
    if (list.isEmpty()) {
        result = null;
    } else {
       ColumnStatisticsDesc csd = new ColumnStatisticsDesc(true, dbName, tableName);
       csd.setCatName(catName);
       result = makeColumnStats(list, csd, 0, engine);
    }
    b.closeAllQueries();
    return result;
  }
```

**Review by belugabehr — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java (hunk 2 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java]: Appreciate that, done. Thank you very much for the review!

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 3]
```diff
@@ -10049,9 +10053,6 @@ public boolean deletePartitionColumnStatistics(String catName, String dbName, St
         }
       }
       ret = commitTransaction();
-    } catch (NoSuchObjectException e) {
-      rollbackTransaction();
-      throw e;
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:10054]: Hey, how is this related to closing query objects?

Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:10054]: It's not related to the query closing, here just to cleanup some unnecessary codes.

**Comment by dengzhhu653** [comment]
Tests passed. Hey @belugabehr @kgyrtkirk could you please take a look? :)

**Comment by dengzhhu653** [comment]
Hi @kgyrtkirk could you please take a look if have secs?

**Comment by dengzhhu653** [comment]
Hi @vihangk1, could you please take a look if have secs?
Thanks!

**Comment by dengzhhu653** [comment]
Hi @pvary, @nrg4878,  could you please take a look if have secs?
Thanks!
Zhihua Deng

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 4]
```diff
@@ -8142,9 +8146,11 @@ private void dropPartitionAllColumnGrantsNoTxn(
       query.declareParameters("java.lang.String t1");
       mSecurityDCList = (List<MDCPrivilege>) query.execute(dcName);
     }
+    try (Query q = query) {
     pm.retrieveAll(mSecurityDCList);
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:8150]: NIT: formatting

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 5]
```diff
@@ -8142,9 +8146,11 @@ private void dropPartitionAllColumnGrantsNoTxn(
       query.declareParameters("java.lang.String t1");
       mSecurityDCList = (List<MDCPrivilege>) query.execute(dcName);
     }
+    try (Query q = query) {
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:8149]: We might have an error during execution. Should we close the query there too?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java [hunk 6]
```diff
@@ -1454,12 +1455,14 @@ public ColumnStatistics getTableStats(final String catName, final String dbName,
...
     List<Object[]> list = Batchable.runBatched(batchSize, colNames, b);
+    final ColumnStatistics result;
     if (list.isEmpty()) {
-      return null;
+      result = null;
+    } else {
+      ColumnStatisticsDesc csd = new ColumnStatisticsDesc(true, dbName, tableName);
+      csd.setCatName(catName);
+      result = makeColumnStats(list, csd, 0, engine);
     }
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:1465]: What happens if there is an exception? Should we close the query there too?

**Comment by pvary** [comment]
Thanks for catching this! 

We should close all the queries, even if there is an exception.
Since Java 8 closing queries in `try-with-resources` would be the best if possible, falling back to `finally` when not possible is the second best.

**Comment by dengzhhu653** [comment]
> Thanks for catching this!
> 
> We should close all the queries, even if there is an exception.
> Since Java 8 closing queries in `try-with-resources` would be the best if possible, falling back to `finally` when not possible is the second best.

Thank you for the review!
Introduces a new QueryWrapper to avoid superfluous exception handing when using `try-with-resources` with the class `Query`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java (hunk 4 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java (hunk 5 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+import javax.jdo.Query;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A wrapper around the Query object to the caller and let the caller release
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java:33]: Why we need this class?

If I understand correctly `Query` implements `AutoCloseable`. Would it be enough to use that?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java [hunk 8]
```diff
@@ -1443,24 +1441,34 @@ public ColumnStatistics getTableStats(final String catName, final String dbName,
         }
         long start = doTrace ? System.nanoTime() : 0;
         Query query = pm.newQuery("javax.jdo.query.SQL", queryText);
-        Object qResult = executeWithArray(query, params, queryText);
-        MetastoreDirectSqlUtils.timingTrace(doTrace, queryText0 + "...)", start, (doTrace ? System.nanoTime() : 0));
-        if (qResult == null) {
-          query.closeAll();
-          return null;
+        try {
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:1444]: Why are you not using the QueryWrapper here?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java [hunk 9]
```diff
@@ -2107,15 +2115,17 @@ private ColumnStatisticsObj prepareCSObjWithAdjustedNDV(Object[] row, int i,
                 makeParams(inputColNames.size()), makeParams(inputPartNames.size()));
             long start = doTrace ? System.nanoTime() : 0;
             Query query = pm.newQuery("javax.jdo.query.SQL", queryText);
-            Object qResult = executeWithArray(query, prepareParams(
-                catName, dbName, tableName, inputPartNames, inputColNames, engine), queryText);
-            MetastoreDirectSqlUtils.timingTrace(doTrace, queryText0, start, (doTrace ? System.nanoTime() : 0));
-            if (qResult == null) {
-              query.closeAll();
-              return Collections.emptyList();
+            try {
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:2118]: I think `addQueryAfterUse` is there for closing the queries. So we might not need them anymore. What do you think?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 10]
```diff
@@ -1554,41 +1553,45 @@ private boolean dropCreationMetadata(String catName, String dbName, String table
     constraintname = constraintname!=null?normalizeIdentifier(constraintname):null;
     List<MConstraint> mConstraints = null;
     List<String> constraintNames = new ArrayList<>();
-    Query query = null;
+    Query queryForConstraintName = null;
+    Query queryForMConstraint = null;
 
     try {
-      query = pm.newQuery("select constraintName from org.apache.hadoop.hive.metastore.model.MConstraint  where "
+      queryForConstraintName = pm.newQuery("select constraintName from org.apache.hadoop.hive.metastore.model.MConstraint  where "
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:1560]: Could we use try-with-resources here as well?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java (hunk 7 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java:33]: Thank you for the review!
The method `close()` of `Query` throws exception,  we should catch the exception explicitly when using use with try-with-resources:
```java
    /**
     * Don't use this method directly; use <code>closeAll()</code> instead. It is intended for use with try-with-resources.
     * @throws Exception if this resource cannot be closed
     */
    void close() throws Exception;
```
The exception would mix with the exception thrown in the `ObjectStore` or `MetaStoreDirectSql`,  make it uneasy to handle on the caller,  as I do not want to break the specified exceptions of the original method.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java (hunk 9 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:2118]: Maybe we should need this,  `Query#execute` may return a lazy result, e.g, the `org.datanucleus.store.rdbms.query.ForwardQueryResult`,  closing the query ealier may result to the exception `rs is closed` when iterating/accessing the result.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java (hunk 8 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:1444]: OK, I will refine this if the `QueryWrapper` seems to be a reasonable choice.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java (hunk 9 above)
Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:2118]: Sounds like a bad practice, but we do not want to solve every problem here 😄

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java (hunk 7 above)
Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java:33]: So the `QueryWrapper` is:

_A wrapper around Query objects where the `close()` method is delegated to the wrapped object's `closeAll()` method. This way the users of the wrapper can use try-with-resources without exception handling_

Disclaimer: I am not a native English speaker, so some more refinement might be needed 😄

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+public class QueryWrapper implements Query {
+
+  private final Query delegate;
+
+  public QueryWrapper(Query query) {
+    requireNonNull(query, "query is null");
+    this.delegate = query;
+  }
+
+  public Query getQuery() {
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java:45]: We do not need this, we might want to remove it

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+    return delegate;
+  }
+
+  @Override
+  public void close() {
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java:50]: I would put a comment here so nobody will try to "fix" this 😄 

Something like:
_Delegates the method to closeAll(), so no exception is thrown_

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java (hunk 7 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java:33]: Thank you for your suggestions, that seems nicer.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java (hunk 11 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java (hunk 12 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/QueryWrapper.java:50]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java (hunk 8 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:1423]: Same as below, closing the query earlier may result to exception with `Query result has been closed`.

**Comment by dengzhhu653** [comment]
Hi @pvary, any comments about the latest changes?  😄 
Thanks,
Zhihua Deng

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java [hunk 13]
```diff
@@ -1850,8 +1819,6 @@ private long partsFoundForPartitions(
           }
           Deadline.checkTimeout();
         }
-      } catch (Exception e) {
-        throwMetaOrRuntimeException(e);
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:1854]: Is this exception handling already happens in the callers, or we end up throwing exceptions which were wrapped before, and will not be wrapped after this change?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java [hunk 14]
```diff
@@ -2125,13 +2080,12 @@ private ColumnStatisticsObj prepareCSObjWithAdjustedNDV(Object[] row, int i,
         }
       }
     };
-    List<Object[]> list = Batchable.runBatched(batchSize, colNames, b);
 
-    List<ColumnStatistics> result = new ArrayList<ColumnStatistics>(
-        Math.min(list.size(), partNames.size()));
+    List<ColumnStatistics> result = new ArrayList<ColumnStatistics>();
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:2084]: nit: I think we can use the `partNames().size()` to initialise the size of the list

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java (hunk 13 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:1854]: The exception handling has aready happened in the caller,  it's allowing to throw MetaException or RuntimeException.
The catch-clause is to catch the exception of releasing query with try-with-resources,  as it throws the  `Exception`, which should be transformed to `MetaException` or `RuntimeException`. This is what the `QueryWrapper` wants to solve,  we do not need this after the change.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java (hunk 14 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java]: done

**Review by pvary — APPROVED** [review]

**Comment by dengzhhu653** [comment]
Hi @pvary, cloud you help merge this? thank you! :)

**Comment by pvary** [comment]
> Hi @pvary, cloud you help merge this? thank you! :)

Ahh... Sorry for forgetting about this review. Could you please ping @belugabehr to check if he is OK with the current solution?

Also we will need to rebase / rerun the tests. I see that theoretically it can be merge without a conflict, but I have had headache several times merging changes with outdated tests which caused problems for the next contributor.

Thanks for keeping this PR alive @dengzhhu653!

**Comment by dengzhhu653** [comment]
> > Hi @pvary, cloud you help merge this? thank you! :)
> 
> Ahh... Sorry for forgetting about this review. Could you please ping @belugabehr to check if he is OK with the current solution?
> 
> Also we will need to rebase / rerun the tests. I see that theoretically it can be merged without a conflict, but I have had headache several times before with merging changes with outdated tests which caused problems for the next contributor.
> 
> Thanks for keeping this PR alive @dengzhhu653!

Thank you very much for the feedback and sorry for the late reply, I will rebase the commits. Hey @belugabehr, cloud you please take another look if have secs?
The issue can be seen in long live shared Metastore client(spark, client pool), the unreleased queries would accumulate to seeing the OOM. The former patch we applied to Hive version 1.2 solved it,  this patch introduces try-with-resource on the query and a sanity check on the master.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java [hunk 15]
```diff
@@ -368,20 +363,22 @@ private void executeNoResult(final String queryText) throws SQLException {
...
+    String queryTextDbSelector= "select "
+        + "\"DB_ID\", \"NAME\", \"DB_LOCATION_URI\", \"DESC\", "
+        + "\"OWNER_NAME\", \"OWNER_TYPE\", \"CTLG_NAME\" , \"CREATE_TIME\", \"DB_MANAGED_LOCATION_URI\", "
+        + "\"TYPE\", \"DATACONNECTOR_NAME\", \"REMOTE_DBNAME\""
+        + "FROM "+ DBS
+        + " where \"NAME\" = ? and \"CTLG_NAME\" = ? ";
+    String queryTextDbParams = "select \"PARAM_KEY\", \"PARAM_VALUE\" "
+        + " from " + DATABASE_PARAMS + " "
+        + " WHERE \"DB_ID\" = ? "
+        + " AND \"PARAM_KEY\" IS NOT NULL";
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:375]: Not related to this PR directly, but there *has* to be a better way of doing this.  This should come from an external resource.

**Comment by belugabehr** [comment]
I'm a little confused here.  What version of JDO is Hive using?

It looks to me like `Query` should already be `Autocloesable`

https://github.com/datanucleus/javax.jdo/blob/fa061bf68e347b8fafca369ab589b7625d8579f3/src/main/java/javax/jdo/Query.java#L68

Thanks.

**Comment by pvary** [comment]
> I'm a little confused here. What version of JDO is Hive using?
> 
> It looks to me like `Query` should already be `Autocloseable`
> 
> https://github.com/datanucleus/javax.jdo/blob/fa061bf68e347b8fafca369ab589b7625d8579f3/src/main/java/javax/jdo/Query.java#L68
> 
> Thanks.

Hehe. I have run the exact same rounds with @dengzhhu653. `Autocloseable`close throws an `Exception`. We do not want to handle this individually, so we need the wrapper 😄

**Review by belugabehr — APPROVED** [review]

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java (hunk 15 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java:375]: Thank you for the review, @belugabehr!  My idea is to make the codes be clean and the same pattern by replacing the standalone `close` to try-with-resource.
